src/libopensc/card-openpgp.c

```diff
@@ -2451,23 +2630,34 @@ pgp_parse_and_set_pubkey_output(sc_card_t *card, u8* data, size_t data_len,
 
 		if (tag == 0x0081) {
 			/* set the output data */
-			if (key_info->rsa.modulus) {
-				memcpy(key_info->rsa.modulus, part, len);
+			if (key_info->u.rsa.modulus) {
+				memcpy(key_info->u.rsa.modulus, part, len);
+			}
+			else {
+				sc_log(card->ctx, "Error: key_info->u.rsa.modulus is NULL");
```

Comment:
I thought long about that function.
If you do not pass the `modulus` and `exponent` as explicit arguments to other functions, may be it's best to malloc() the necessary memory and copy those elements in any case.
@hongquan What was your intention here?

Reply:
Honestly, I am not sure, if I understand the issue here.

Do you mean in case the card response is different than requested? Like, the modulus length differs from what was inserted in the private key template? In this case I would prefer an error message instead.
Would it be sufficient to add
```
if (key_info->rsa.modulus && len == key_info->rsa.modulus_len){
```
as the value stored in key_info->rsa.modulus_len should be the size of the allocated memory (like it is done for the ec_point)?

Reply:
Sorry, @marschap, I'm not understand what you mean here. Could you please give an example?

Reply:
@hongquan 
I was referring to the whole code section starting in line 2627
```
                if (tag == 0x0081) {
                        /* set the output data */
                        if (key_info->u.rsa.modulus) {
                                memcpy(key_info->u.rsa.modulus, part, len);
                        }
                        else {
                                sc_log(card->ctx, "Error: key_info->u.rsa.modulus is NULL");
                        }
                        /* always set output for modulus_len */
                        key_info->u.rsa.modulus_len = len*8;
                }
```
I am wondering why the modulus parsed (in variable `part`) is only copied to the key_info structure if the `key_info->u.rsa.modulus` pointer is not NULL.
* Shouldn't it be copied in any case?
  I.e. shouldn't the memory be allocated it it is NULL?
  Especially with the change now that the explicit `modulus` variable & parameter to `pgp_calculate_and_store_fingerprint()` got (IMHO correctly) removed by @alex-nitrokey 's patch.
* Furthermore, shouldn't it be checked that the size allocated for `key_info->u.rsa.modulus` is large enough to keep the contents of the parsed modulus.
 E.g. when going from a 2k modulus to a 4k modulus, `key_info->u.rsa.modulus` may perhaps be allocated but too small.

@alex-nitrokey I think your code catches at least one potential error case and reports error on it.
The issues above can be left for a later commit after @honquan clarified.

Reply:
I didn't think that `key_info->u.rsa.modulus` can be less than enough, because it was `malloc` with the size of expected key length.

But I'm happy that you add more checking that I might miss.

Reply:
I agree with @marschap , this whole section needs length checking! The length is obtained from `sc_asn1_read_tag`, which may initialize `len` to something beyond the length of the buffer, because you're not erroring on `SC_ERROR_ASN1_END_OF_CONTENTS`.

Reply:
I mostly reused previous code. I trusted this code more than myself. I try to stop this behaviour :wink:

I tried to add some (hopefully) proper checks now. Please have a look at it.

`SC_ERROR_ASN1_END_OF_CONTENTS` is imho included in
```
	r = sc_asn1_read_tag((const u8**)&part, data_len - (in - data), &cla, &tag, &len);
 	if (part == NULL)
 		r = SC_ERROR_ASN1_OBJECT_NOT_FOUND;
 	LOG_TEST_RET(card->ctx, r, "Unexpected end of contents.");
```